### ISSUE TITLE
Provide a pagination widget by default

### ISSUE THREAD

**Issue by anderscheow** [flutter/flutter#24141]
Currently the possible ways to achieve pagination are using ScrollView or ListView.Builder. But need to consider about loading indicator and retry mechanism. Although it's completely possible to create such widget by ourself but it's great if Flutter team able to build this widget.

**Comment by AbdulRahmanAlHamali** [comment]
Hello,

I have developed such a package:

https://pub.dartlang.org/packages/flutter_pagewise

Would that help?

**Comment by JohannesKaufmann** [comment]
There are actually at least two more packages implementing this:
- [lazy_load_scrollview](https://pub.dartlang.org/packages/lazy_load_scrollview)
- [incrementally_loading_listview](https://pub.dartlang.org/packages/incrementally_loading_listview)

But Infinite scroll is such a crucial feature for any app (it's one of the ways to make apps [addictive](https://www.bbc.com/news/technology-44640959)) that I think it should be part of the core or at least officially made by flutter. The current plugins don't fit into the "sliver world" that nicely: If you have an existing `CustomScrollView` with a `SliverList` and a `SliverGrid` it's not that easy to implement any of the three plugins easily (and it's not clear which of the plugins is the best one).

**Comment by AbdulRahmanAlHamali** [comment]
@JohannesKaufmann you are right regarding the Sliver problem. I noticed it just yesterday when using the plugin in one of my projects. So I will try to update today to make it work with Slivers, and will let you know

**Comment by AbdulRahmanAlHamali** [comment]
@JohannesKaufmann [there you go](https://pub.dartlang.org/packages/flutter_pagewise)

**Comment by JohannesKaufmann** [comment]
@AbdulRahmanAlHamali The problem I talked about is something different: You have a paginated list of sections where each section has a title and ids.

```json
{
  "sections": [
    {
       "title": "Title 1",
       "ids": [1, 2, 3, 4]
    },
    {
       "title": "Title 2",
       "ids": [1, 2, 3, 4]
    },
    // ...many sections -> paginated
  ]
}
```

Currently I have `CustomScrollView` containing: The "title" in a `SliverToBoxAdapter` and the "ids" in a  `SliverGrid`. 

While your recent additions (wow that was quick, btw) are definetely a valid use case it unfortunately does not resolve my issue. 

I don't know either how to design an api for that. The react community has the [render-prop](https://reactjs.org/docs/render-props.html) pattern (aka: builder parameter) which I personally really like. All the logic is done by the plugin but the visualls are done by the user.
So maybe pass the `scroll_controller`, `data` (like isLoading) and `actions` (like retry) down to the widget that implements it (which is written by the user of the plugin).

**Comment by AbdulRahmanAlHamali** [comment]
@JohannesKaufmann how about you structure your view as follows, wouldn't that solve the problem?

```dart
CustomScrollView(
  slivers: [
    PagewiseSliverList(
      pageSize: 10, // for example,
      pageFuture: this._pageFuture, // some function that fetches the page
      itemBuilder: (context, entry, index) {
        return Column(
          children: [
            Text(entry['title']),
            ListView(
              primary: false,
              shrinkWrap: true,
              children: entry['ids'].map((id) => Text(id)).toList()
            )
          ]
        );
      }
    )
  ]
)
```

**Comment by anderscheow** [comment]
May I know what is "Silver" thingy?

**Comment by zoechi** [comment]
@anderscheow you might want to watch https://www.youtube.com/watch?v=Mz3kHQxBjGg (see also https://www.youtube.com/results?search_query=flutter+slivers)

**Comment by Gopalakrishnan-V** [comment]
Any update on this? @zoechi @timsneath

**Comment by timsneath** [comment]
It's not on our immediate roadmap to build this into the framework (see the definition of the [Future milestone](https://github.com/flutter/flutter/milestone/10)). There are some good packages that implement pagination, and you might find that you want to customize this experience anyway. There are some good articles that describe how to implement this, for example [this one](https://medium.com/@jun.chenying/flutter-tutorial-part-5-listview-pagination-scroll-up-to-load-more-ed132f6a06be). 

(Note, there's also a [PaginatedDataTableClass](https://api.flutter.dev/flutter/material/PaginatedDataTable-class.html), although that solves a different class of pagination issues to the one described above.)

Our general focus is on improving the core framework to make it easier for others to build this kind of capability, rather than filling out the package ecosystem, which is why this is marked as Future.
